### PR TITLE
Add support for logging the model checkpoints to MLFlowLogger

### PR DESCRIPTION
**Pull request by AlessioQuercia**
## What does this PR do?
This PR adds a log_model parameter to MLFlowLogger as the one in WandbLogger (#9138).

The log_model parameter can be set to:
- "all": checkpoints are logged during training.
- True: checkpoints are logged at the end of training, except when 
`:paramref: '~pytorch_lightning.callbacks.model_checkpoint.ModelCheckpoint.save_top_k' == "-1"` which also logs every checkpoint during training.
- False: no checkpoint is logged.

In particular, the function behavior is not exactly the same as in WandbLogger, since MLFlowClient().log_artifact takes different parameters as input:
- In the WandbLogger, the artifact is logged together with a metadata dictionary and some aliases.
- In the MLFlowLogger, it is not possible to log metadata and aliases together with the artifact, therefore (to keep consistency) a temporary folder is created, metadata.yml and aliases.txt are created inside of it, then the temporary folder is logged on mlflow together with the checkpoint artifact in the model/checkpoints/CHECKPOINT_NAME/ folder.

### Does your PR introduce any breaking changes? If yes, please list them.
None.

## Before submitting
- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Coding this small PR was funny 😃

### CODE REVIEW

File: src/pytorch_lightning/loggers/mlflow.py [hunk 1]
```diff
@@ -127,6 +139,7 @@ def __init__(
         tracking_uri: Optional[str] = os.getenv("MLFLOW_TRACKING_URI"),
         tags: Optional[Dict[str, Any]] = None,
         save_dir: Optional[str] = "./mlruns",
+        log_model: Optional[bool] = False,
```
Comment (awaelchli) [inline src/pytorch_lightning/loggers/mlflow.py:142]: The type here needs to reflect the different types of choices listed in the docs above. So I suggest to go for 

```suggestion
        log_model: Literal[True, False, "all"] = False,
```

File: src/pytorch_lightning/loggers/mlflow.py [hunk 2]
```diff
@@ -108,6 +113,13 @@ def any_lightning_module_function_or_hook(self):
         save_dir: A path to a local directory where the MLflow runs get saved.
             Defaults to `./mlflow` if `tracking_uri` is not provided.
             Has no effect if `tracking_uri` is provided.
+        log_model: Log checkpoints created by :class:`~pytorch_lightning.callbacks.Checkpoint`
```
Comment (awaelchli) [inline src/pytorch_lightning/loggers/mlflow.py:116]:
```suggestion
        log_model: Log checkpoints created by :class:`~pytorch_lightning.callbacks.model_checkpoint.ModelCheckpoint`
```

File: src/pytorch_lightning/CHANGELOG.md [hunk 3]
```diff
@@ -19,6 +19,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Added
 
+- Added `log_model` parameter to `MLFlowLogger` ([#9187](https://github.com/PyTorchLightning/pytorch-lightning/pull/9187))
```
Comment (awaelchli) [inline src/pytorch_lightning/CHANGELOG.md:22]: This feature would likely go into 1.9, so let's add it into the "unreleased" section above.

File: src/pytorch_lightning/loggers/mlflow.py [hunk 4]
```diff
@@ -108,6 +113,13 @@ def any_lightning_module_function_or_hook(self):
...
+            as MLFlow artifacts.
+            * if ``log_model == 'all'``, checkpoints are logged during training.
```
Comment (awaelchli) [inline src/pytorch_lightning/loggers/mlflow.py:118]:
```suggestion
            as MLFlow artifacts.
            
            * if ``log_model == 'all'``, checkpoints are logged during training.
```
The empty lines are needed so bullet points get rendered correctly in the docs.

File: src/pytorch_lightning/loggers/mlflow.py [hunk 5]
```diff
@@ -108,6 +113,13 @@ def any_lightning_module_function_or_hook(self):
...
+            * if ``log_model == True``, checkpoints are logged at the end of training, except when
+              :paramref:`~pytorch_lightning.callbacks.Checkpoint.save_top_k` ``== -1``
+              which also logs every checkpoint during training.
+            * if ``log_model == False`` (default), no checkpoint is logged.
         prefix: A string to put at the beginning of metric keys.
```
Comment (awaelchli) [inline src/pytorch_lightning/loggers/mlflow.py:123]:
```suggestion
            * if ``log_model == False`` (default), no checkpoint is logged.

        prefix: A string to put at the beginning of metric keys.
```

File: tests/tests_pytorch/loggers/test_mlflow.py [hunk 6]
```diff
@@ -277,3 +278,45 @@ def test_mlflow_logger_finalize_when_exception(*_):
...
+def test_mlflow_log_model(client, mlflow, tmpdir, log_model):
+    """Test that the logger creates the folders and files in the right place."""
+
+    # Reset mocks
+    client.return_value.log_artifact.reset_mock()
+    client.return_value.log_artifacts.reset_mock()
+    # Get model, logger, trainer and train
+    model = BoringModel()
+    logger = MLFlowLogger("test", save_dir=tmpdir, log_model=log_model)
+    checkpoint_callback = ModelCheckpoint(dirpath=tmpdir / "model_checkpoints")
```
Comment (awaelchli) [inline tests/tests_pytorch/loggers/test_mlflow.py:295]: Is there a specific reason why you chose to create a checkpoint callback instance with this particular path? `Trainer(enable_checkpointing=True)` would also work (default) right?

**Review by awaelchli — APPROVED** [review]

**Comment by awaelchli** [comment]
@carmocca or @Borda or @justusschock this should be ready for re-review.

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:

**Review by justusschock — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
